proposals/p3720.md

```diff
@@ -0,0 +1,853 @@
+interfaces. Also allow classes to `extend api` other classes, for forwarding use
+cases.
+
+## Problem
+
+What happens when binding is performed between an object instance and a member
+of its type? We'd like to define the semantics in a way that is simple,
+orthogonal, supports the use cases from C++, allows users to express their
+intent in code in a natural and predictable way that is consistent with other
+Carbon constructs, and is consistent with Carbon's goals.
+
+Consider a class with a method and a field:
+
+```carbon
+class C {
+  fn M[self: Self]();
+  var f: i32;
+}
+var x: C = {.f = 2};
+```
+
+The expressions `C.M` and `C.f` correspond roughly to
+[C++ pointers to members](https://en.cppreference.com/w/cpp/language/pointer#Pointers_to_members).
+They may be used to access the members of `x` using Carbon's
+[compound member syntax](/docs/design/expressions/member_access.md), as in
+`x.(C.M)` or `x.(C.f)`. What is their type? Can they be passed to a function
+separately from the instance of `C` to bind with it?
+
+The other issue is how we clearly delineate the `self` associated with a method
+signature as separate from the `self` of function values.
+
+## Background
+
+Member access has been specified in two proposals:
+
+-   [Proposal #989: Member access expressions](https://github.com/carbon-language/carbon-lang/pull/989)
+-   [Proposal #2360: Types are values of type `type`](https://github.com/carbon-language/carbon-lang/pull/2360)
+
+The results of these proposals is recorded in the
+["qualified names and member access" design document](/docs/design/expressions/member_access.md).
+Notably, there is the process of
+[instance binding](/docs/design/expressions/member_access.md#instance-binding),
+that can convert a method into a bound method. This is described as an
+uncustomizable process, with the members of classes being non-first-class names.
+
+With
+[proposal #3646: Tuples and tuple indexing](https://github.com/carbon-language/carbon-lang/pull/3646),
+tuple indexing also uses the member-access syntax, except with numeric names for
+the fields.
+
+The currently accepted proposals for functions, most notably
+[Proposal #2875: Functions, function types, and function calls](https://github.com/carbon-language/carbon-lang/pull/2875),
+don't support all of the different function signatures for the `Call` interface.
+For example, it does not support `addr self` or explicit compile-time
+parameters. That is out of scope of this proposal, and will be addressed
+separately, and means that `addr self` methods won't be considered here. The
+difference between functions and methods, however, is in scope.
+
+## Proposal
+
+We propose that Carbon defines the compound member access operator, specifically
+`x.(y)`, to be defined as in terms of rewrites to invoking an interface method,
+like other operators. There are two different interfaces used, depending on
+whether `x` is a value expression or a reference expression:
```

Comment:
Maybe I'm missing something about how these interfaces work, so I'd like to ask a question that's been bouncing in my head; suppose I have:

```
// Class to target with calls.
class D {}

// Declare a class which requires C -> D conversions.
class C {
  fn Foo[self: D]();
  fn Bar[addr self: D*]();

  // Apologies if I get my syntax wrong here: trying to support RefBind of C -> D.
  extend impl as RefBind(D) where .Result = D {
    fn Op[self: Self](p: C*) -> D* {
      return p as D*;
    }
  }
}

fn Run() {
  // I assume this succeeds because ValueBind isn't checked at this point;
  // although, maybe I should ask, maybe it's worth diagnosing that
  // the Foo function requires a missing ValueBind?
  var c: C;
  // I assume this uses RefBind.
  c.Bar();
  // This fails due to lack of ValueBind, right? Could it succeed using RefBind?
  c.Foo();
}
```

Reply:
I wonder if we should do more for interfaces that compose separate interfaces, but I expect that's out of scope here.